### ISSUE TITLE
dockerng.running: Unable to bind mount a file if it already exists in the image

### ISSUE THREAD

**Issue by kkoppel** [saltstack/salt#29183]
Hi!

This happens when trying to create a container with dockerng and bind a file from the host when a file with the same name already exists inside the image:

```
# salt-call state.single dockerng.running name=test image='debian:jessie' binds='["/etc/passwd:/etc/passwd"]'
local:
----------
          ID: test
    Function: dockerng.running
      Result: False
     Comment: Failed to create new container: Error 500: file exists at /var/lib/docker/aufs/mnt/9eeaf4a5f2ebcb6489d1b10dd47e09c091969f4d88f0c29ddd87cac0889a3470/etc/passwd, can't create volume there
     Started: 08:41:53.489985
    Duration: 781.056 ms
     Changes:
```

The above works fine if the destination path inside the image does not exist.

FYI, I found a bug with the exact same error reported for docker-compose (https://github.com/docker/compose/issues/1326) and I suspect that this error happens due to the way Salt is using docker-py, because the following python code works fine for me:

```
import docker
cli = docker.Client(base_url='unix:///var/run/docker.sock', version='auto')
container = cli.create_container(name='test', image='debian:jessie', command='/bin/sleep 600', volumes=['/etc/passwd'], host_config=docker.Client.create_host_config(cli, binds=['/etc/passwd:/etc/passwd'], network_mode='bridge'))
cli.start(container=container.get('Id'))
```

And here are the Salt and Docker version infos:

```
# salt-call --versions-report
Salt Version:
           Salt: 2015.8.1

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: Not Installed
           Mako: 1.0.0
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.2
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: debian 8.2 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: debian 8.2

# docker -v
Docker version 1.6.2, build 7c8fca2
```

**Comment by ticosax** [comment]
Hi,
I tried to reproduce your issue, but failed to do so.
I'm using `docker==1.9.1` and `docker-py==1.5.0`

```
$ salt-call state.single dockerng.running name=test image='debian:jessie' binds='["/etc/passwd:/etc/passwd"]'
[INFO    ] Loading fresh modules for state activity
[INFO    ] Running state [test] at time 10:13:40.804292
[INFO    ] Executing state dockerng.running for test
[WARNING ] /usr/lib/python2.7/dist-packages/salt/modules/data.py:120: DeprecationWarning: Support for 'getval' has been deprecated and will be removed in Salt Boron. Please use 'get' instead.

[INFO    ] {'state': {'new': 'running', 'old': None}, 'added': {'Time_Elapsed': 0.19596290588378906, 'Name': 'test', u'Id': u'f8c37503bc119021fbc9103def53d17529afa65a12f9717403d422a39f89cf07', u'Warnings': None}}
[INFO    ] Completed state [test] at time 10:13:41.405121
local:
----------
          ID: test
    Function: dockerng.running
      Result: True
     Comment: Container 'test' changed state.. Container 'test' was created.
     Started: 10:13:40.804292
    Duration: 600.829 ms
     Changes:   
              ----------
              added:
                  ----------
                  Id:
                      f8c37503bc119021fbc9103def53d17529afa65a12f9717403d422a39f89cf07
                  Name:
                      test
                  Time_Elapsed:
                      0.195962905884
                  Warnings:
                      None
              state:
                  ----------
                  new:
                      running
                  old:
                      None

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 600.829 ms

```
```
salt-call --versions-report
Salt Version:
           Salt: 2015.8.2-53-g345206b
 
Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: 0.9.1
         PyYAML: 3.10
          PyZMQ: 14.0.1
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
           RAET: Not Installed
        Tornado: 4.3
            ZMQ: 4.0.4
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.3.0
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 0.3.6
          smmap: 0.8.2
        timelib: Not Installed
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-46-generic
         system: Ubuntu 14.04 trusty
 
```

**Comment by kkoppel** [comment]
Thanks for checking!
I forgot to mention that I also have docker-py v1.5.0.

I ran some additional tests in a clean test environment and found that if I only upgrade to Salt v2015.8.2, then I get a different error message:
```
Failed to create new container: Unable to perform create_container: create_container() got an unexpected keyword argument 'binds'
```
If I now also upgrade to Docker v1.9.1, then yes, it works fine.

However, since Docker version v1.6.2 is the latest Docker version currently available from the official Debian Jessie package repositories (backports), then this version is likely to be widely used. Would be nice if support for this Docker version could be fixed in Salt as well.

**Comment by jfindlay** [comment]
@kkoppel, thanks for the report.  You are welcome to send in a pull request against 2015.8 that modifies the salt behavior if an older version of docker is detected.

**Comment by terminalmage** [comment]
This is almost certainly fixed by 2017.7.0. Marking as closed, please reply and we can re-visit it if there is still a problem.
